api-select.go

```diff
@@ -0,0 +1,491 @@
+		return nil, err
+	}
+
+	selectReqBytes, err := xml.Marshal(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	urlValues := make(url.Values)
+	urlValues.Set("select", "")
+	urlValues.Set("select-type", "2")
+
+	// Execute POST on bucket/object.
+	resp, err := c.executeMethod(ctx, "POST", requestMetadata{
+		bucketName:       bucketName,
+		objectName:       objectName,
+		queryValues:      urlValues,
+		customHeader:     opts.Header(),
+		contentMD5Base64: sumMD5Base64(selectReqBytes),
+		contentSHA256Hex: sum256Hex(selectReqBytes),
+		contentBody:      bytes.NewReader(selectReqBytes),
+		contentLength:    int64(len(selectReqBytes)),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, httpRespToErrorResponse(resp, bucketName, "")
+	}
+
+	streamer := &SelectResults{
+		resp:     resp,
+		stats:    &StatsMessage{},
+		progress: &ProgressMessage{},
+	}
+	streamer.start()
+	return streamer, nil
+}
+
+// Close - closes the underlying response body and the stream reader.
+func (s *SelectResults) Close() error {
+	defer closeResponse(s.resp)
+	return s.pipeReader.Close()
+}
+
+// Read - is a reader compatible implementation for SelectObjectContent records.
+func (s *SelectResults) Read(b []byte) (n int, err error) {
+	return s.pipeReader.Read(b)
+}
+
+// Stats - information about a request's stats when processing is complete.
+func (s *SelectResults) Stats() *StatsMessage {
+	return s.stats
+}
+
+// Progress - information about the progress of a request.
+func (s *SelectResults) Progress() *ProgressMessage {
+	return s.progress
+}
+
+// start is the main function that decodes the large byte array into
+// several events that are sent through the eventstream.
+func (s *SelectResults) start() {
+	var pipeWriter *io.PipeWriter
+	s.pipeReader, pipeWriter = io.Pipe()
+	go func() {
+		for {
+			var prelude preludeInfo
+			var headers *headerInfo
+			var err error
+			var payload []byte
+
+			// Create CRC code
+			crc := crc32.New(crc32.IEEETable)
+			crcReader := io.TeeReader(s.resp.Body, crc)
+
+			// Extract the prelude(12 bytes) into a struct to extract relevant information.
+			prelude, err = processPrelude(crcReader, crc)
+			if err != nil {
+				pipeWriter.CloseWithError(err)
+				closeResponse(s.resp)
+				return
+			}
+
+			// Extract the headers(variable bytes) into a struct to extract relevant information
+			if prelude.headerLen > 0 {
+				headers, err = extractHeader(io.LimitReader(crcReader, int64(prelude.headerLen)))
+				if err != nil {
+					pipeWriter.CloseWithError(err)
+					closeResponse(s.resp)
+					return
+				}
+			}
+
+			// Calculate the length of the payload so that the appropriate amount of
+			// bytes can be read
+			payloadLen := calcPayloadLen(prelude)
+			// Checks that there is a payload and if there is, then extracts it.
+			if payloadLen > 0 {
+				payload, err = extractPayload(io.LimitReader(crcReader, int64(payloadLen)), payloadLen)
```

Comment:
This will take up some memory in the client application if the payload is large (potentially 4GB), and increase latency while the payload gets buffered inside the library - it can be avoided by writing to the client directly (perhaps with a pipe).

Reply:
yes we can write directly @donatello